### PR TITLE
ServerIcon: Fix race condition and add additional safety check

### PR DESCRIPTION
**Pull request by kristofbolyai**
Fixes: 
```
---- Minecraft Crash Report ----
// Don't do that.

Time: 5/23/22, 5:25 PM
Description: Unexpected error

java.lang.NullPointerException: Cannot invoke "com.mojang.blaze3d.platform.NativeImage.m_84982_()" because "this.f_117977_" is null
	at net.minecraft.client.renderer.texture.DynamicTexture.m_117992_(DynamicTexture.java:21) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:classloading}
	at com.mojang.blaze3d.systems.RenderSystem.m_69884_(RenderSystem.java:211) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:classloading}
	at com.mojang.blaze3d.systems.RenderSystem.m_69495_(RenderSystem.java:198) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:classloading}
	at com.mojang.blaze3d.platform.Window.m_85435_(Window.java:333) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:classloading,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91383_(Minecraft.java:1066) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:wynntils.mixins.json:MinecraftMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:663) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:wynntils.mixins.json:MinecraftMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.main.Main.main(Main.java:205) ~[client-1.18.2-20220404.173914-srg.jar%2353!/:?] {re:classloading,pl:runtimedistcleaner:A}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?] {}
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?] {}
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?] {}
	at net.minecraftforge.fml.loading.targets.CommonClientLaunchHandler.lambda$launchService$0(CommonClientLaunchHandler.java:45) ~[fmlloader-1.18.2-40.0.40.jar%2316!/:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:37) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:106) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:77) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:149) [bootstraplauncher-1.0.0.jar:?] {}
```

#

First issue to be actually found by (or at least pointed out :sweat_smile:) [Qodana](https://github.com/JetBrains/qodana-action) :tada:. I am looking forward to adding it to our Github CI as soon as #162 is fixed. Until then, I will try to make a nice default configuration file for it.  
image
image

### CODE REVIEW

File: common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java [hunk 1]
```diff
@@ -112,8 +116,8 @@ private synchronized void loadServerIcon(ResourceLocation destination) {
         Validate.validState(nativeImage.getHeight() == 64, "Must be 64 pixels high");
 
         synchronized (this) {
-            serverIconLocation = destination;
             Minecraft.getInstance().getTextureManager().register(destination, new DynamicTexture(nativeImage));
+            serverIconLocation = destination;
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:120]: Why does this fix the race condition?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:120]: Ah, I didn't fully read the code that renders the serverIcon, thought it was about setting the icon location before registering it. Anyways, even if this does not improve anything, it certainly won't cause harm.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:120]: Perhaps, the DynamicTexture constructor does some rendering task that is the culprit.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:120]: But it is also weird, the crash log in PR description hints that `nativeImage` is null when DynamicTexture is constructed. How can that even happen?

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:120]: The crash log doesn't mention `ServerIcon` at all.

**Comment by magicus** [comment]
I have a hard time figure out this PR. (Partly because I have a hard time figuring out the code it modifies). Let's start with the basics:

* I've never seen this crash. Is this something that has happened more than once for you? Can you repeat it by provoking it somehow.

* Adding "volatile" seems scary. If we have a race between threads this is unlikely to be a fix, it might only make the issue less prevalent.

* I'm not sure what Qodanas insight is here. If we pass null and that is not allowed, then we'd most likely crash consistently.

**Comment by kristofbolyai** [comment]
2am coding strikes again. I reviewed my code, and I cam agree this doesn't fix anything obvious (although it doesn't hurt anything). As for the crash, this is existent, but it is random, sometimes very common, sometimes rare.

I will close this PR and maybe dive deeper into why it happens.

**Comment by magicus** [comment]
I don't really understand the ServerIcon code. I suspect we are mis-using Vanilla code, somehow...

**Pull request closed without merging**